Improve Goroutines metric calls in parallelizer.Until

macsko opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Incrementing and decrementing `Goroutines` metric in `parallelizer.Until` gives a noticeable overhead, especially for InterPodAffinity and PodTopologySpread plugins that use it to parallelize some operations in filtering and scoring. Before, these operations were performed more times than enough - per single work piece, not per single worker (goroutine). This PR changes the behavior in `parallelizer.Until` by doing `Add` and `Dec` once per goroutine. 

This solution improves the performance of most affinity and topology spread scenarios by up to 10%.

scheduler_perf benchmarks showing improvement (local):
Benchmark | Before [pods/s avg] | After [pods/s avg]
--- | --- | ---
SchedulingPodAntiAffinity/5000Nodes_2000Pods_QueueingHintsEnabled | 161 | 181
SchedulingPodAffinity/5000Nodes_5000Pods_QueueingHintsEnabled | 106 | 114
SchedulingPreferredPodAffinity/5000Nodes_5000Pods_QueueingHintsEnabled | 315 | 335
SchedulingPreferredPodAntiAffinity/5000Nodes_5000Pods_QueueingHintsEnabled | 316 | 336
SchedulingNodeAffinity/5000Nodes_10000Pods_QueueingHintsEnabled | 595 | 633
MixedSchedulingBasePod/5000Nodes_5000Pods_QueueingHintsEnabled | 323 | 429
SchedulingRequiredPodAffinityWithNSSelector/5000Nodes_2000Pods_QueueingHintsEnabled | 106 | 112
SchedulingPreferredAffinityWithNSSelector/5000Nodes_5000Pods_QueueingHintsEnabled | 296 | 315
TopologySpreading/5000Nodes_5000Pods_QueueingHintsEnabled | 263 | 283

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

macsko commented:
/cc @sanposhiho @dom4ha

haosdent commented on pkg/scheduler/framework/parallelize/parallelism.go:
```diff
@@ -51,15 +51,27 @@ func chunkSizeFor(n, parallelism int) int {
 		goroutinesMetric.Dec()
 	}
+	chunkSize := chunkSizeFor(pieces, p.parallelism)
+	goroutinesMetric.Add(float64(workersForChunkSize(p.parallelism, pieces, chunkSize)))
```
Should we have a comment here to mention why we write it this way is to ensure performance?

haosdent commented:
/lgtm

sanposhiho commented on pkg/scheduler/framework/parallelize/parallelism.go:
```diff
@@ -51,15 +51,29 @@ func chunkSizeFor(n, parallelism int) int {
+// workersForChunkSize returns number of workers (goroutines)
+// that will be created in workqueue.ParallelizeUntil
+// for given parallelism, pieces and chunkSize values.
+func workersForChunkSize(parallelism, pieces, chunkSize int) int {
```
nit:
```suggestion
func numWorkersForChunkSize(parallelism, pieces, chunkSize int) int {
```

sanposhiho commented on pkg/scheduler/framework/parallelize/parallelism.go:
```diff
@@ -51,15 +51,29 @@ func chunkSizeFor(n, parallelism int) int {
 	withMetrics := func(piece int) {
-		goroutinesMetric.Inc()
 		doWorkPiece(piece)
 		goroutinesMetric.Dec()
```
According to the implementation of `ParallelizeUntil`, won't `Dec()` called more than the worker num calculated from `workersForChunkSize()`?
- `workersForChunkSize()` equals to the number of goroutines. 
  - IIUC, it returns the same number as `workers` [here](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L75).
- `goroutinesMetric.Dec()` could be called multiple times within goroutines.
  - https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L90

haosdent replied:
It seems it is equal to chunks instead of workers.

macsko replied:
Ouch, that's right. So now, I'm not sure if my change is possible easily. Before, `Inc()` and `Dec()` were called `pieces` number of times and were capped by number of `workers`. We can't simply do `Add(pieces)`, because we would register much more goroutines than we have. 

I'm wondering if we couldn't do batch `Add(-workers)` (similar to introduced `Add(workers)`) instead of decrementing per each work piece. It won't be fully accurate, but still will give us the most important semantic, which is finding leaked goroutines.

macsko replied:
Please check the actual implementation

haosdent commented on pkg/scheduler/framework/parallelize/parallelism.go:
```diff
@@ -51,15 +51,28 @@ func chunkSizeFor(n, parallelism int) int {
-	}
+	// Calling single Add with workers' count is more efficient than calling Inc or Dec per each work piece.
+	// This approach improves performance of some plugins (affinity, topology spreading).
+	goroutinesMetric.Add(float64(workers))
```
I try to understand the code again (Sorry, my math is not good 😂), but it always seems to be equal to pieces. I write a program to verify

```go
package main

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
)

func main() {
	// Test cases to verify
	testCases := []struct {
		workers   int
		pieces    int
		chunkSize int
	}{
		{2, 5, 2},   // Example from our discussion
		{3, 6, 2},   // Another example
		{2, 10, 3},  // Uneven chunks
		{5, 3, 1},   // More workers than pieces
		{1, 100, 7}, // Large number of pieces
	}

	for _, tc := range testCases {
		var callCount atomic.Int64

		// Create a function that counts calls
		doWork := func(piece int) {
			callCount.Add(1)
		}

		// Run the parallelizer
		ParallelizeUntil(context.Background(), tc.workers, tc.pieces, doWork, WithChunkSize(tc.chunkSize))

		fmt.Printf("Workers: %d, Pieces: %d, ChunkSize: %d => Actual calls: %d (Expected: %d)\n",
			tc.workers, tc.pieces, tc.chunkSize, callCount.Load(), tc.pieces)
	}
}

// Copy of the original ParallelizeUntil and supporting code for testing
func ParallelizeUntil(ctx context.Context, workers, pieces int, doWorkPiece func(piece int), opts ...Options) {
	if pieces == 0 {
		return
	}
	o := options{}
	for _, opt := range opts {
		opt(&o)
	}
	chunkSize := o.chunkSize
	if chunkSize < 1 {
		chunkSize = 1
	}

	chunks := ceilDiv(pieces, chunkSize)
	toProcess := make(chan int, chunks)
	for i := 0; i < chunks; i++ {
		toProcess <- i
	}
	close(toProcess)

	var stop <-chan struct{}
	if ctx != nil {
		stop = ctx.Done()
	}
	if chunks < workers {
		workers = chunks
	}
	wg := sync.WaitGroup{}
	wg.Add(workers)
	for i := 0; i < workers; i++ {
		go func() {
			defer wg.Done()
			for chunk := range toProcess {
				start := chunk * chunkSize
				end := start + chunkSize
				if end > pieces {
					end = pieces
				}
				for p := start; p < end; p++ {
					select {
					case <-stop:
						return
					default:
						doWorkPiece(p)
					}
				}
			}
		}()
	}
	wg.Wait()
}

type options struct {
	chunkSize int
}

type Options func(*options)

func WithChunkSize(c int) func(*options) {
	return func(o *options) {
		o.chunkSize = c
	}
}

func ceilDiv(a, b int) int {
	return (a + b - 1) / b
}
```

Output is

```
Workers: 2, Pieces: 5, ChunkSize: 2 => Actual calls: 5 (Expected: 5)
Workers: 3, Pieces: 6, ChunkSize: 2 => Actual calls: 6 (Expected: 6)
Workers: 2, Pieces: 10, ChunkSize: 3 => Actual calls: 10 (Expected: 10)
Workers: 5, Pieces: 3, ChunkSize: 1 => Actual calls: 3 (Expected: 3)
Workers: 1, Pieces: 100, ChunkSize: 7 => Actual calls: 100 (Expected: 100)
```

haosdent replied:
BTW, it is weird that `goroutinesMetric` didn't equal the number of actual workers although it named `goroutines metric`

macsko replied:
We have a few different values in `parallelizer.Until`:
- parallelism (named `workers` in `ParallelizeUntil`) - how many goroutines (at most) we want to dispatch
- pieces - how many times we want to call the `doWorkPiece` function
- chunkSize - how many work pieces are processed at once within a goroutine
- chunks - into how many chunks the pieces are split (`ceilDiv(pieces, chunkSize)`)
- actual workers - is equal to `min(parallelism, chunks)` - that's the return value of `numWorkersForChunkSize`

The `actual workers` is the number of goroutines dispatched in `ParallelizeUntil`. In your code indeed `callCount.Load() == tc.piece`, because we run the `doWorkPiece` for each work piece. It doesn't mean that the `Goroutines` metric is wrong - it is capped at `actual workers` value here, as it is incremented before each work piece and decremented right after. We cannot have more than `actual workers` work pieces running at once, so it (almost) correctly counts the number of running goroutines. Almost, because we increment and decrement the metric multiple times within a single goroutine, what doesn't make sense.

sanposhiho replied:
It's obvious that `callCount` equals `pieces`, right? What are you trying to say here?

haosdent replied:
> Almost, because we increment and decrement the metric multiple times within a single goroutine, what doesn't make sense.

Yep, this is where I feel the original metric is a bit weird. By right it should be increased outside of `for chunk := range toProcess {`

> It's obvious that callCount equals pieces, right? What are you trying to say here?

I was thinking that may adjust the original code to make it record metrics accurately. 

```
	withMetrics := func(piece int) {
		doWorkPiece(piece)
		goroutinesMetric.Dec()
	}
	goroutinesMetric.Add(float64(pieces))
```

sanposhiho reviewed (approved):
/approve

dom4ha commented:
What kind of improvement this change was supposed to bringing? My first impression is that the code complication might be not worth the benefit, that's why I'm asking about the potential gain.

dom4ha commented:
The code after changes looks much more straightforward now (we do increment and decrement at the same place), so no concerns about readability anymore.

I still don't like a bit that we need to do some calculations in our package assuming the logic of the other package, so I'd consider extending paralelizer with optional metic to increment and decrement (if it was provided) on goroutine creation/completion:

func WithMetric(m *metrics.GaugeVec) func(*options) {
	return func(o *options) {
		o.metric = m
	}
}
WDYT?

aojea commented:
> This solution improves the performance of most affinity and topology spread scenarios up to 10%.

is common providing benchmarks that proof the claims when talking about performance gains, although I leave to the maintainers of the codebase the decision

macsko commented:
I implemented @dom4ha's idea. @sanposhiho PTAL.

> is common providing benchmarks that proof the claims when talking about performance gains

I added them to the description of PR.

dom4ha reviewed (commented):
You still need to set the metric I think

macsko commented:
> You still need to set the metric I think

I don't understand. Metric is passed to the `workqueue.ParallelizeUntil` and it's working.

dom4ha reviewed (commented):
/lgtm

> I don't understand. Metric is passed to the workqueue.ParallelizeUntil and it's working.

Ah, it's at the end

sanposhiho reviewed (approved):
/lgtm
/approve

For the scheduling part.

sanposhiho commented:
/retest

macsko commented:
/assign @aojea for `staging/src/k8s.io/client-go` changes

fedebongio commented:
/assign @aojea 
/triage accepted

macsko commented:
Ping @aojea

aojea commented on staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:
```diff
@@ -76,6 +85,12 @@ func ParallelizeUntil(ctx context.Context, workers, pieces int, doWorkPiece DoWo
 			defer utilruntime.HandleCrash()
 			defer wg.Done()
+
+			if o.workerMetric != nil {
+				o.workerMetric.Inc()
+				defer o.workerMetric.Dec()
```
you are not measuring the same now, this measure each time a worker is created, process all pieces and then exits.
IIUIC, previous metrics measure everytime a piece is processed.

It seems the performance improvement is related to the fact that the metric is different now

macsko replied:
I would say that the previous implementation measured the same thing, but in a slightly incorrect way, because it incremented and decremented for each work item, which caused the metric to reach at most the number of goroutines (workers), but for a short time it could be lower (when it was between processing two pieces by one worker). 

See [comment](https://github.com/kubernetes/kubernetes/pull/128999#discussion_r1861880803) with explanation.

aojea commented on staging/src/k8s.io/client-go/util/workqueue/parallelizer.go:
```diff
@@ -41,6 +42,14 @@ func WithChunkSize(c int) func(*options) {
 	}
 }
 
+// WithWorkerMetric allows to set a metric that will be incremented
+// at the beginning of a worker (goroutine) and decremented at its end.
+func WithWorkerMetric(m GaugeMetric) func(*options) {
+	return func(o *options) {
+		o.workerMetric = m
+	}
+}
+
```
This metrics seems that was added to check if there are leaks https://github.com/kubernetes/kubernetes/pull/112003/files#r957348631

Something that leaks goroutines is buggy, that can not happen, adding a metrics to detect bugs in production does not look great, we are adding goleaks functionality to all integration test and some unit test, if the metrics only servers that purpose and you are worried about performance, metrics are global and expensive, then it is better to completly remove the metrics and add unit and/or integratino tests that detect gorouting leaks, see some examples here
```
staging/src/k8s.io/client-go/tools/record/event_test.go:        "go.uber.org/goleak"
staging/src/k8s.io/client-go/tools/record/event_test.go:        goleak.VerifyNone(t)
staging/src/k8s.io/client-go/tools/record/main_test.go: "go.uber.org/goleak"
staging/src/k8s.io/client-go/tools/record/main_test.go: goleak.VerifyTestMain(m)
```

The `ParallelizeUntil` method is using a `WaitGroup` so it can not leak goroutines, it can only hang ...

sanposhiho replied:
> it can not leak goroutines, it can only hang ...

Yes, so what if one goroutine is taking long time somehow?
Also, we use the metric in other plain goroutines started with `go func() {...}()` as well, which can be leak technically. 
I obviously agree we shouldn't have a goroutine leak in the first place though, the leak can happen due to user's implementation that is run within the framework, I mean, custom scheduler plugins.

aojea replied:
yeah, that  is exactly the friction point, this is moving an implementation detail of the scheduling plugins to a public API ... can we agree that this option is very bespoke and only useful for this specific case?
Is there any other way we don't leak this scheduling problem into this functionality?

@macsko if you don't need this metric at all what if we try to make it optional in the `Parallelizer` inside the framework instead of here?

alculquicondor replied:
Well, there is a PRR question that goes along the lines of: "how do we know if this feature is not working". This is what this metric is trying to answer.
This is valid not just for external plugins, but also for upstream.

But I'm open to making the metric recording opt-out. That is, once a feature is known to be stable, remove the metric recording for it's parallelization routines.

sanposhiho replied:
Or, can we just bring back the implementation before @dom4ha 's suggestion? 
https://github.com/kubernetes/kubernetes/pull/128999#issuecomment-2506081675

That would allow us to improve the throughput before making changes within client-go side.

macsko replied:
> Or, can we just bring back the implementation before @dom4ha 's suggestion?

Yeah, I think that would be the best solution if we don't want to change `client-go/util`, keep the metric and keep the performance gain. I'll revert the PR to those changes.

macsko committed:
Improve Goroutines metric calls in parallelizer.Until

macsko commented:
I reverted the changes to the state before https://github.com/kubernetes/kubernetes/pull/128999#issuecomment-2506081675. @sanposhiho @dom4ha PTAL for ltgm.

sanposhiho reviewed (commented):
/approve

Leave `/lgtm` to @dom4ha.

macsko commented:
/retest
